src/Razor/src/Microsoft.AspNetCore.Razor.ExternalAccess.RoslynWorkspace/RazorWorkspaceListenerBase.cs

```diff
@@ -0,0 +1,246 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT license. See License.txt in the project root for license information.
+
+using System.Collections.Immutable;
+using Microsoft.AspNetCore.Razor.Utilities;
+using Microsoft.CodeAnalysis;
+using Microsoft.Extensions.Logging;
+
+namespace Microsoft.AspNetCore.Razor.ExternalAccess.RoslynWorkspace;
+
+public abstract class RazorWorkspaceListenerBase : IDisposable
+{
+    private static readonly TimeSpan s_debounceTime = TimeSpan.FromMilliseconds(500);
+    private readonly CancellationTokenSource _disposeTokenSource = new();
+
+    private readonly ILogger _logger;
+    private readonly AsyncBatchingWorkQueue<Work> _workQueue;
+
+    // Use an immutable dictionary for ImmutableInterlocked operations. The value isn't checked, just
+    // the existance of the key so work is only done for projects with dynamic files.
+    private ImmutableDictionary<ProjectId, bool> _projectsWithDynamicFile = ImmutableDictionary<ProjectId, bool>.Empty;
+
+    private Stream? _stream;
+    private Workspace? _workspace;
+
+    internal record Work(ProjectId ProjectId);
+    internal record UpdateWork(ProjectId ProjectId) : Work(ProjectId);
+    internal record RemovalWork(ProjectId ProjectId, string IntermediateOutputPath) : Work(ProjectId);
+
+    public RazorWorkspaceListenerBase(ILogger logger)
+    {
+        _logger = logger;
+        _workQueue = new(TimeSpan.FromMilliseconds(500), UpdateCurrentProjectsAsync, EqualityComparer<Work>.Default, _disposeTokenSource.Token);
+    }
+
+    void IDisposable.Dispose()
+    {
+        if (_workspace is not null)
+        {
+            _workspace.WorkspaceChanged -= Workspace_WorkspaceChanged;
+        }
+
+        if (_disposeTokenSource.IsCancellationRequested)
+        {
+            return;
+        }
+
+        _disposeTokenSource.Cancel();
+        _disposeTokenSource.Dispose();
+
+        _stream?.Dispose();
+        _stream = null;
+    }
+
+    public void NotifyDynamicFile(ProjectId projectId)
+    {
+        // Since there is no "un-notify" API to indicate that callers no longer care about a project, it's entirely
+        // possible that by the time we get notified, a project might have been removed from the workspace. Whilst
+        // that wouldn't cause any issues we may as well avoid creating a task scheduler.
+        if (_workspace is null || !_workspace.CurrentSolution.ContainsProject(projectId))
+        {
+            return;
+        }
+
+        ImmutableInterlocked.GetOrAdd(ref _projectsWithDynamicFile, projectId, static (_) => true);
+
+        // Schedule a task, in case adding a dynamic file is the last thing that happens
+        _logger.LogTrace("{projectId} scheduling task due to dynamic file", projectId);
+        _workQueue.AddWork(new UpdateWork(projectId));
```

Comment:
At first I was a bit worried what would happen if the project was removed while still handling this; I guess we'll see the removal after this event handler, and as long as the work queue doesn't reorder the remove operation then you're good.

Reply:
We use the `_workspace.CurrentSolution` and double check the project is still present. I believe that should suffice to make sure, and on the other end if we receive a remove that we never knew about we just ignore.